mockrxandroidble/src/main/java/com/polidea/rxandroidble/mockrxandroidble/RxBleConnectionMock.java

```diff
@@ -0,0 +1,198 @@
+package com.polidea.rxandroidble.mockrxandroidble;
+
+import android.bluetooth.BluetoothGattCharacteristic;
+import android.bluetooth.BluetoothGattDescriptor;
+import android.support.annotation.NonNull;
+import android.support.v4.util.Pair;
+
+import com.polidea.rxandroidble.RxBleConnection;
+import com.polidea.rxandroidble.RxBleDeviceServices;
+import com.polidea.rxandroidble.exceptions.BleCannotSetCharacteristicNotificationException;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+
+import rx.Observable;
+import rx.Subscriber;
+
+import static android.bluetooth.BluetoothGattDescriptor.DISABLE_NOTIFICATION_VALUE;
+import static android.bluetooth.BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE;
+import static rx.Observable.error;
+
+class RxBleConnectionMock implements RxBleConnection {
+
+    private final HashMap<UUID, Observable<Observable<byte[]>>> notificationObservableMap = new HashMap<>();
+    private RxBleDeviceServices rxBleDeviceServices;
+    private Integer rssid;
+    private Map<UUID, Observable<byte[]>> characteristicNotificationSources;
+
+
+    public RxBleConnectionMock(RxBleDeviceServices rxBleDeviceServices,
+                               Integer rssid,
+                               Map<UUID, Observable<byte[]>> characteristicNotificationSources) {
+        this.rxBleDeviceServices = rxBleDeviceServices;
+        this.rssid = rssid;
+        this.characteristicNotificationSources = characteristicNotificationSources;
+    }
+
+    @Override
+    public Observable<RxBleDeviceServices> discoverServices() {
+        return Observable.just(rxBleDeviceServices);
+    }
+
+    public Observable<BluetoothGattCharacteristic> getCharacteristic(@NonNull UUID characteristicUuid) {
+        return discoverServices()
+                .flatMap(rxBleDeviceServices -> rxBleDeviceServices.getCharacteristic(characteristicUuid));
+    }
+
+    @Override
+    public Observable<byte[]> readCharacteristic(@NonNull UUID characteristicUuid) {
+        return getCharacteristic(characteristicUuid).map(BluetoothGattCharacteristic::getValue);
+    }
+
+    @Override
+    public Observable<byte[]> readDescriptor(UUID serviceUuid, UUID characteristicUuid, UUID descriptorUuid) {
+        return discoverServices()
+                .flatMap(rxBleDeviceServices -> rxBleDeviceServices.getDescriptor(serviceUuid, characteristicUuid, descriptorUuid))
+                .map(BluetoothGattDescriptor::getValue);
+    }
+
+    @Override
+    public Observable<Integer> readRssi() {
+        return Observable.just(rssid);
+    }
+
+    @Override
+    public Observable<Observable<byte[]>> setupNotification(@NonNull UUID characteristicUuid) {
+        synchronized (notificationObservableMap) {
```

Comment:
We don't need synchronization for mocking purposes. Multiple observables can be created safely here.